File: framec_tests/golang/simple_handler_calls/simple_handler_calls.go
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

package simple_handler_calls

import "golang/framelang"


func NewSimpleHandlerCalls() SimpleHandlerCalls {
    m := &simpleHandlerCallsStruct{}
    
    // Validate interfaces
    var _ SimpleHandlerCalls = m
    
    
    // Create and intialize start state compartment.
    m._compartment_ = NewSimpleHandlerCallsCompartment(SimpleHandlerCallsState_Init)
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type SimpleHandlerCallsState uint

const (
    SimpleHandlerCallsState_Init SimpleHandlerCallsState = iota
    SimpleHandlerCallsState_A
    SimpleHandlerCallsState_B
)

type SimpleHandlerCalls interface {
    A() 
    B() 
    C() 
    D() 
    E() 
}

type simpleHandlerCallsStruct struct {
    _compartment_ *SimpleHandlerCallsCompartment
    _nextCompartment_ *SimpleHandlerCallsCompartment
}

//===================== Interface Block ===================//

func (m *simpleHandlerCallsStruct) A()  {
    e := framelang.FrameEvent{Msg:"A"}
    m._mux_(&e)
}

func (m *simpleHandlerCallsStruct) B()  {
    e := framelang.FrameEvent{Msg:"B"}
    m._mux_(&e)
}

func (m *simpleHandlerCallsStruct) C()  {
    e := framelang.FrameEvent{Msg:"C"}
    m._mux_(&e)
}

func (m *simpleHandlerCallsStruct) D()  {
    e := framelang.FrameEvent{Msg:"D"}
    m._mux_(&e)
}

func (m *simpleHandlerCallsStruct) E()  {
    e := framelang.FrameEvent{Msg:"E"}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *simpleHandlerCallsStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case SimpleHandlerCallsState_Init:
        m._SimpleHandlerCallsState_Init_(e)
    case SimpleHandlerCallsState_A:
        m._SimpleHandlerCallsState_A_(e)
    case SimpleHandlerCallsState_B:
        m._SimpleHandlerCallsState_B_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *simpleHandlerCallsStruct) _SimpleHandlerCallsState_Init_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "A":
        compartment := NewSimpleHandlerCallsCompartment(SimpleHandlerCallsState_A)
        m._transition_(compartment)
        
        return
    case "B":
        compartment := NewSimpleHandlerCallsCompartment(SimpleHandlerCallsState_B)
        m._transition_(compartment)
        
        return
    case "C":
        m.A()
        return
        
        return
    case "D":
        m.B()
        return
        compartment := NewSimpleHandlerCallsCompartment(SimpleHandlerCallsState_A)
        m._transition_(compartment)
        
        return
    case "E":
        m.D()
        return
        m.C()
        return
        
        return
    }
}

func (m *simpleHandlerCallsStruct) _SimpleHandlerCallsState_A_(e *framelang.FrameEvent) {
    switch e.Msg {
    }
}

func (m *simpleHandlerCallsStruct) _SimpleHandlerCallsState_B_(e *framelang.FrameEvent) {
    switch e.Msg {
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *simpleHandlerCallsStruct) _transition_(compartment *SimpleHandlerCallsCompartment) {
    m._nextCompartment_ = compartment
}

func (m *simpleHandlerCallsStruct) _do_transition_(nextCompartment *SimpleHandlerCallsCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}


//=============== Compartment ==============//

type SimpleHandlerCallsCompartment struct {
    State SimpleHandlerCallsState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewSimpleHandlerCallsCompartment(state SimpleHandlerCallsState) *SimpleHandlerCallsCompartment {
    c := &SimpleHandlerCallsCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}

File: framec_tests/python/src/simple_handler_calls/simple_handler_calls.py
#Emitted from framec_v0.11.0



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class SimpleHandlerCalls:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = SimpleHandlerCallsCompartment('__simplehandlercalls_state_Init', next_compartment)
        self.__compartment = next_compartment
        self.__next_compartment = None
        
        # Initialize domain
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def A(self,):
        __e = FrameEvent("A",None)
        self.__kernel(__e)
    
    def B(self,):
        __e = FrameEvent("B",None)
        self.__kernel(__e)
    
    def C(self,):
        __e = FrameEvent("C",None)
        self.__kernel(__e)
    
    def D(self,):
        __e = FrameEvent("D",None)
        self.__kernel(__e)
    
    def E(self,):
        __e = FrameEvent("E",None)
        self.__kernel(__e)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $Init
    
    def __simplehandlercalls_state_Init(self, __e, compartment):
        if __e._message == "A":
            next_compartment = None
            next_compartment = SimpleHandlerCallsCompartment('__simplehandlercalls_state_A', next_compartment)
            
            self.__transition(next_compartment)
            return
        elif __e._message == "B":
            next_compartment = None
            next_compartment = SimpleHandlerCallsCompartment('__simplehandlercalls_state_B', next_compartment)
            
            self.__transition(next_compartment)
            return
        elif __e._message == "C":
            self.A()
            return
            return
        elif __e._message == "D":
            self.B()
            return
            next_compartment = None
            next_compartment = SimpleHandlerCallsCompartment('__simplehandlercalls_state_A', next_compartment)
            
            self.__transition(next_compartment)
            return
        elif __e._message == "E":
            self.D()
            return
            self.C()
            return
            return
    
    # ----------------------------------------
    # $A
    
    def __simplehandlercalls_state_A(self, __e, compartment):
        pass
        
    
    # ----------------------------------------
    # $B
    
    def __simplehandlercalls_state_B(self, __e, compartment):
        pass
        
    
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__simplehandlercalls_state_Init':
            self.__simplehandlercalls_state_Init(__e, self.__compartment)
        elif self.__compartment.state == '__simplehandlercalls_state_A':
            self.__simplehandlercalls_state_A(__e, self.__compartment)
        elif self.__compartment.state == '__simplehandlercalls_state_B':
            self.__simplehandlercalls_state_B(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class SimpleHandlerCallsCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    